relay-server/src/actors/project_buffer.rs

```diff
@@ -290,55 +338,26 @@ impl BufferService {
-                        "INSERT INTO envelopes (own_key, sampling_key, envelope) ",
-                    );
-
-                    // Flatten all the envelopes
-                    let envelopes = buf.into_iter().flat_map(|(k, vals)| {
-                        vals.into_iter()
-                            .map(move |v| (k, v.into_envelope().to_vec()))
-                    });
-
-                    // Since we have 3 variables we have to bind, we devide the SQLite limit by 3
-                    // here to prepare the chnunks which will be preparing the batch inserts.
-                    let mut envelopes =
-                        stream::iter(envelopes).chunks(SQLITE_LIMIT_VARIABLE_NUMBER / 3);
-                    while let Some(chunk) = envelopes.next().await {
-                        query_builder.push_values(chunk.into_iter(), |mut b, v| match v.1 {
-                            Ok(envelope_bytes) => {
-                                b.push_bind(v.0.own_key.to_string())
-                                    .push_bind(v.0.sampling_key.to_string())
-                                    .push_bind(envelope_bytes);
-                            }
-                            Err(err) => {
-                                relay_log::error!(
-                                    "failed to serialize the envelope: {}",
-                                    LogError(&err)
-                                )
-                            }
-                        });
-
-                        let query = query_builder.build();
-                        if let Err(err) = query.execute(&db).await {
-                            relay_log::error!(
-                                "failed to buffer envelopes to disk: {}",
-                                LogError(&err)
-                            )
-                        }
-                        // Reset the builder to initial state set by `QueryBuilder::new` function,
-                        // so it can be reused for another chunk.
-                        query_builder.reset();
-                    }
-                });
+                    // A RESERVED lock will be acquired when the first INSERT, UPDATE, or DELETE statement is executed.
+                    // Once we started spooling our in-memory buffer, the dequeue from the disk
+                    // will be blocked till this operations either fails, or commits.
+                    sqlx::query("BEGIN").execute(&mut connection).await?;
+
+                    Self::do_spool(&mut connection, buf).await?;
+
+                    // The SQL command "COMMIT" does not actually commit the changes to disk. It just turns autocommit back on.
+                    // Then, at the conclusion of the command, the regular autocommit logic takes over and causes the actual commit to disk to occur.
+                    // TODO: re-try failed commit?
+                    sqlx::query("COMMIT").execute(&mut connection).await?;
```

Comment:
This explicit open/close transaction has to go away now. And we will rely on the pool to do the right thing, since we handle all the messages sequentially now